### PR TITLE
Story/10568 ticket validation allow list

### PR DESCRIPTION
**Pull request by 30mar**
## Description
fetch the allow list and perform related checks

## Link to Jira
(https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-10568)](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-10568)

### CODE REVIEW

File: src/xcode/ENA/ENA/Source/Services/TicketValidation/AllowList/AllowListService.swift [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+	// MARK: - Private
+
+	private let restServiceProvider: RestServiceProviding
+	private let store: Store
+	
+	// MARK: - Internal
+
+	func fetchAllowList() {
+		let resource = AllowListResource()
+		restServiceProvider.load(resource) { result in
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Services/TicketValidation/AllowList/AllowListService.swift:26]: self inside used might leak without [weak self] used in the closure.

File: src/xcode/ENA/ENA/Source/Services/TicketValidation/AllowList/AllowListService.swift [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+					ValidationServiceAllowlistEntry(
+						serviceProvider: $0.serviceProvider,
+						hostname: $0.hostname,
+						fingerprint256: $0.fingerprint256.base64EncodedString()
+					)
+				})
+				self.store.ticketValidationAllowList = TicketValidationAllowList(
+					validationServiceAllowList: validationServiceAllowlist,
+					serviceProviderAllowList: serviceProviderAllowlist
+				)
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Services/TicketValidation/AllowList/AllowListService.swift:42]: I don't see why another cache is require here?
You will get the cached data from the CachedRestService for free.

File: src/xcode/ENA/ENA/Source/Workers/Store/SecureStore.swift [hunk 3]
```diff
@@ -595,7 +595,23 @@ extension SecureStore: CoronaTestStoringLegacy {
...
+				serviceProviderAllowList: []
+			)
+		}
+		set { kvStore["ticketValidationAllowList"] = newValue }
+	}
+	
+	var serviceProviderAllowlist: [Data] {
+		get { kvStore["serviceProviderAllowlist"] as [Data]? ?? [] }
+		set { kvStore["serviceProviderAllowlist"] = newValue }
+	}
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Workers/Store/SecureStore.swift:614]: won't this double the cache? It's already stored inside the cached rest service.

**Review by KaiTeuber — CHANGES_REQUESTED** [review]

File: src/xcode/ENA/ENA/Source/Workers/Store/SecureStore.swift (hunk 3 above)
Comment (30mar) [reply src/xcode/ENA/ENA/Source/Workers/Store/SecureStore.swift:614]: I wasnt sure about this, i thought we only store the eTag, how can can i access the cached stuff ?

File: src/xcode/ENA/ENA/Source/Services/TicketValidation/AllowList/AllowListService.swift [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+		certificateChain: [Data], // can be construct by: Data(base64Encoded: x509String)
+		allowlist: [ValidationServiceAllowlistEntry]
+	) -> Result<Void, AllowListError> {
+		guard let leafCertificate = certificateChain.first else {
+			Log.debug("Certificate chain should include at least one certificate", log: .ticketValidationAllowList)
+			return .failure(.CERT_PIN_MISMATCH)
+		}
+		// Find requiredFingerprints: the requiredFingerprints shall be set by mapping each entry in allowlist to their fingerprint256 attribute.
+
+		let requiredFingerprints = allowlist.map({
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Services/TicketValidation/AllowList/AllowListService.swift:66]:
```suggestion
		let requiredFingerPrints = allowlist.map({
```

File: src/xcode/ENA/ENA/Source/Services/TicketValidation/AllowList/AllowListService.swift [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+			$0.fingerprint256
+		})
+		
+		// Compare fingerprints: if the SHA-256 fingerprints of leafCertificate is not included in requiredFingerprints, the operation shall abort with error code CERT_PIN_MISMATCH.
+		let leafFingerPrint = leafCertificate.sha256().base64EncodedString()
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Services/TicketValidation/AllowList/AllowListService.swift:71]: Fingerprint is one word

```suggestion
		let leafFingerprint = leafCertificate.sha256().base64EncodedString()
```

File: src/xcode/ENA/ENA/Source/Services/TicketValidation/AllowList/AllowListService.swift [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+		if requiredFingerprints.contains(where: {
+			$0 == leafFingerPrint
+		}) {
+			Log.debug("fingerprints found", log: .ticketValidationAllowList)
+		} else {
+			return .failure(.CERT_PIN_MISMATCH)
+		}
+		
+		let requiredHostnames: [String] = allowlist.compactMap({
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Services/TicketValidation/AllowList/AllowListService.swift:80]:
```suggestion
		let requiredHostNames: [String] = allowlist.compactMap({
```

File: src/xcode/ENA/ENA/Source/Services/TicketValidation/AllowList/AllowListService.swift [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+			$0.fingerprint256 == leafFingerPrint ? $0.hostname : nil
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Services/TicketValidation/AllowList/AllowListService.swift:81]:
```suggestion
			$0.fingerprint256 == leafFingerprint ? $0.hostname : nil
```

File: src/xcode/ENA/ENA/Source/Services/TicketValidation/AllowList/AllowListService.swift [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+		})
+		if requiredHostnames.contains(where: {
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Services/TicketValidation/AllowList/AllowListService.swift:83]:
```suggestion
		if requiredHostNames.contains(where: {
```

File: src/xcode/ENA/ENA/Source/Services/TicketValidation/AllowList/AllowListService.swift (hunk 5 above)
Comment (30mar) [reply src/xcode/ENA/ENA/Source/Services/TicketValidation/AllowList/AllowListService.swift:71]: The naming is correct as we are following the techspec convention here, please check TrustEvaluation.swift as well

File: src/xcode/ENA/ENA/Source/Services/TicketValidation/AllowList/AllowListService.swift (hunk 7 above)
Comment (30mar) [reply src/xcode/ENA/ENA/Source/Services/TicketValidation/AllowList/AllowListService.swift:81]: The naming is correct as we are following the techspec convention here, please check TrustEvaluation.swift as well

File: src/xcode/ENA/ENA/Source/Services/TicketValidation/AllowList/AllowListService.swift (hunk 5 above)
Comment (naveeddotio) [reply src/xcode/ENA/ENA/Source/Services/TicketValidation/AllowList/AllowListService.swift:71]: Then we should follow the same convention in line 72. 

It should be then requiredFingerPrints

File: src/xcode/ENA/ENA/Source/Services/TicketValidation/AllowList/AllowListService.swift [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+		}
+		// Find requiredFingerprints: the requiredFingerprints shall be set by mapping each entry in allowlist to their fingerprint256 attribute.
+
+		let requiredFingerprints = allowlist.map({
+			$0.fingerprint256
+		})
+		
+		// Compare fingerprints: if the SHA-256 fingerprints of leafCertificate is not included in requiredFingerprints, the operation shall abort with error code CERT_PIN_MISMATCH.
+		let leafFingerPrint = leafCertificate.sha256().base64EncodedString()
+		if requiredFingerprints.contains(where: {
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Services/TicketValidation/AllowList/AllowListService.swift:72]:
```suggestion
		if requiredFingerPrints.contains(where: {
```

**Review by naveeddotio — APPROVED** [review]

File: src/xcode/ENA/ENA/Source/Services/TicketValidation/AllowList/AllowListService.swift [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+
+	private let restServiceProvider: RestServiceProviding
+	
+	// MARK: - Internal
+
+	@OpenCombine.Published var allowlist: TicketValidationAllowList?
+
+	func fetchAllowList() {
+		let resource = AllowListResource()
+		restServiceProvider.load(resource) { result in
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Services/TicketValidation/AllowList/AllowListService.swift:29]:
```suggestion
		restServiceProvider.load(resource) { [weak self] result in
```

File: src/xcode/ENA/ENA/Source/Services/TicketValidation/AllowList/AllowListService.swift [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+			switch result {
+			case .success(let allowListProtoBuf):
+				self.allowlist = allowListProtoBuf.allowlist
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Services/TicketValidation/AllowList/AllowListService.swift:32]:
```suggestion
				self?.allowlist = allowListProtoBuf.allowlist
```

**Review by KaiTeuber — APPROVED** [review]

**Pull request merged**